File: java/iroha_android/src/test/java/org/hyperledger/iroha/android/offline/OfflineQrStreamTest.java
package org.hyperledger.iroha.android.offline;

import java.util.List;

public final class OfflineQrStreamTest {

  private OfflineQrStreamTest() {}

  public static void main(final String[] args) {
    roundTripPayload();
    recoversMissingChunk();
    rejectsBadChecksum();
    textCodecRoundTrip();
    sakuraStormPlaybackSkinMatchesPreset();
    sakuraStormScanSessionPresetRecoversDroppedFrame();
    System.out.println("[IrohaAndroid] OfflineQrStreamTest passed.");
  }

  private static void roundTripPayload() {
    final byte[] payload = makePayload(1024);
    final List<byte[]> frames =
        OfflineQrStream.Encoder.encodeFrameBytes(
            payload, OfflineQrStream.PayloadKind.UNSPECIFIED, new OfflineQrStream.Options(200, 0));
    final OfflineQrStream.Decoder decoder = new OfflineQrStream.Decoder();
    OfflineQrStream.DecodeResult result = null;
    for (byte[] frame : frames) {
      result = decoder.ingest(frame);
    }
    assertNotNull(result, "decode result");
    assertTrue(result.isComplete(), "roundtrip incomplete");
    assertArrayEquals(payload, result.payload(), "roundtrip payload mismatch");
  }

  private static void recoversMissingChunk() {
    final byte[] payload = makePayload(900);
    final List<OfflineQrStream.Frame> frames =
        OfflineQrStream.Encoder.encodeFrames(
            payload,
            OfflineQrStream.PayloadKind.OFFLINE_SPEND_RECEIPT,
            new OfflineQrStream.Options(180, 3));
    OfflineQrStream.Frame header = null;
    int droppedIndex = -1;
    for (OfflineQrStream.Frame frame : frames) {
      if (frame.kind() == OfflineQrStream.FrameKind.HEADER) {
        header = frame;
        break;
      }
    }
    assertNotNull(header, "header");
    final OfflineQrStream.Decoder decoder = new OfflineQrStream.Decoder();
    decoder.ingest(header.encode());
    OfflineQrStream.DecodeResult result = null;
    for (OfflineQrStream.Frame frame : frames) {
      if (frame.kind() == OfflineQrStream.FrameKind.DATA && frame.index() == 1) {
        droppedIndex = frame.index();
        continue;
      }
      if (frame.kind() != OfflineQrStream.FrameKind.HEADER) {
        result = decoder.ingest(frame.encode());
      }
    }
    assertTrue(droppedIndex == 1, "dropped index mismatch");
    assertNotNull(result, "decode result");
    assertTrue(result.isComplete(), "parity recovery incomplete");
    assertArrayEquals(payload, result.payload(), "parity payload mismatch");
  }

  private static void rejectsBadChecksum() {
    final byte[] payload = makePayload(400);
    final List<byte[]> frames =
        OfflineQrStream.Encoder.encodeFrameBytes(
            payload, OfflineQrStream.PayloadKind.UNSPECIFIED, new OfflineQrStream.Options());
    final byte[] corrupted = frames.get(0).clone();
    corrupted[corrupted.length - 1] ^= (byte) 0x11;
    final OfflineQrStream.Decoder decoder = new OfflineQrStream.Decoder();
    boolean threw = false;
    try {
      decoder.ingest(corrupted);
    } catch (IllegalArgumentException error) {
      threw = true;
    }
    assertTrue(threw, "checksum mismatch should throw");
  }

  private static void textCodecRoundTrip() {
    final byte[] payload = makePayload(128);
    final String encoded =
        OfflineQrStream.TextCodec.encode(payload, OfflineQrStream.FrameEncoding.BASE64);
    final byte[] decoded =
        OfflineQrStream.TextCodec.decode(encoded, OfflineQrStream.FrameEncoding.BASE64);
    assertArrayEquals(payload, decoded, "text codec payload mismatch");
  }

  private static void sakuraStormPlaybackSkinMatchesPreset() {
    final OfflineQrStream.PlaybackSkin skin = OfflineQrStream.SAKURA_STORM_SKIN;
    assertTrue("sakura-storm".equals(skin.name), "storm skin name mismatch");
    assertTrue(skin.frameRate == 12.0, "storm frameRate mismatch");
    assertTrue(skin.petalDriftSpeed == 0.6, "storm drift mismatch");
    assertTrue(skin.progressOverlayAlpha == 0.34, "storm overlay mismatch");
    assertTrue(skin.theme.backgroundStart.red == 0.05, "storm backgroundStart mismatch");
    assertTrue(skin.theme.backgroundEnd.blue == 0.04, "storm backgroundEnd mismatch");
  }

  private static void sakuraStormScanSessionPresetRecoversDroppedFrame() {
    final byte[] payload = makePayload(6 * 1024);
    final List<OfflineQrStream.Frame> frames =
        OfflineQrStream.Encoder.encodeFrames(
            payload,
            OfflineQrStream.PayloadKind.OFFLINE_SPEND_RECEIPT,
            new OfflineQrStream.Options(336, 4));

    OfflineQrStream.Frame header = null;
    for (OfflineQrStream.Frame frame : frames) {
      if (frame.kind() == OfflineQrStream.FrameKind.HEADER) {
        header = frame;
        break;
      }
    }
    assertNotNull(header, "header");

    final OfflineQrStream.ScanSession session = new OfflineQrStream.ScanSession();
    session.ingest(header.encode());

    OfflineQrStream.DecodeResult result = null;
    for (OfflineQrStream.Frame frame : frames) {
      if (frame.kind() == OfflineQrStream.FrameKind.DATA && frame.index() == 1) {
        continue;
      }
      if (frame.kind() == OfflineQrStream.FrameKind.HEADER) {
        continue;
      }
      result = session.ingest(frame.encode());
    }

    assertNotNull(result, "storm decode result");
    assertTrue(result.isComplete(), "storm scan session incomplete");
    assertArrayEquals(payload, result.payload(), "storm scan payload mismatch");
    assertTrue(result.recoveredChunks() == 1, "storm recovered chunk mismatch");
  }

  private static byte[] makePayload(final int length) {
    final byte[] payload = new byte[length];
    for (int i = 0; i < payload.length; i++) {
      payload[i] = (byte) ((i * 31 + 7) & 0xFF);
    }
    return payload;
  }

  private static void assertNotNull(final Object value, final String label) {
    if (value == null) {
      throw new AssertionError(label + " must not be null");
    }
  }

  private static void assertTrue(final boolean condition, final String message) {
    if (!condition) {
      throw new AssertionError(message);
    }
  }

  private static void assertArrayEquals(
      final byte[] expected, final byte[] actual, final String message) {
    if (expected == null && actual == null) {
      return;
    }
    if (expected == null || actual == null || expected.length != actual.length) {
      throw new AssertionError(message);
    }
    for (int i = 0; i < expected.length; i++) {
      if (expected[i] != actual[i]) {
        throw new AssertionError(message);
      }
    }
  }
}

File: javascript/iroha_js/test/offlineQrStream.test.js
import assert from "node:assert/strict";
import fs from "node:fs";
import test from "node:test";

import {
  OfflineQrStreamDecoder,
  OfflineQrStreamEncoder,
  OfflineQrStreamFrameKind,
  OfflineQrStreamFrameEncoding,
  OfflineQrStreamScanSession,
  encodeQrFrameText,
  decodeQrFrameText,
  scanQrStreamFrames,
  sakuraStormQrStreamTheme,
  sakuraStormQrStreamSkin,
} from "../src/offlineQrStream.js";

function buildPayload(size = 1200) {
  const buffer = Buffer.alloc(size);
  for (let i = 0; i < buffer.length; i += 1) {
    buffer[i] = (i * 31 + 7) % 256;
  }
  return buffer;
}

test("offline qr stream round-trips payloads", () => {
  const payload = buildPayload();
  const frames = OfflineQrStreamEncoder.encodeFrameBytes(payload, {
    chunkSize: 200,
  });
  const decoder = new OfflineQrStreamDecoder();
  let result = null;
  for (const frame of frames) {
    result = decoder.ingest(frame);
  }
  assert.ok(result);
  assert.equal(result.isComplete, true);
  assert.deepEqual(result.payload, payload);
});

test("offline qr stream recovers a missing frame with parity", () => {
  const payload = buildPayload(900);
  const frames = OfflineQrStreamEncoder.encodeFrames(payload, {
    chunkSize: 180,
    parityGroup: 3,
  });
  const header = frames.find((frame) => frame.kind === OfflineQrStreamFrameKind.header);
  const dataFrames = frames.filter((frame) => frame.kind === OfflineQrStreamFrameKind.data);
  const parityFrames = frames.filter((frame) => frame.kind === OfflineQrStreamFrameKind.parity);
  assert.ok(header);
  assert.equal(dataFrames.length > 0, true);
  assert.equal(parityFrames.length > 0, true);
  const dropped = dataFrames[1];
  const decoder = new OfflineQrStreamDecoder();
  decoder.ingest(header.encode());
  let result = null;
  for (const frame of dataFrames) {
    if (frame !== dropped) {
      result = decoder.ingest(frame.encode());
    }
  }
  for (const frame of parityFrames) {
    result = decoder.ingest(frame.encode());
  }
  assert.ok(result);
  assert.equal(result.isComplete, true);
  assert.deepEqual(result.payload, payload);
});

test("offline qr stream rejects frames with invalid checksum", () => {
  const payload = buildPayload(400);
  const frames = OfflineQrStreamEncoder.encodeFrameBytes(payload);
  const frame = Buffer.from(frames[0]);
  frame[frame.length - 1] ^= 0x22;
  const decoder = new OfflineQrStreamDecoder();
  assert.throws(() => decoder.ingest(frame), /checksum/i);
});

test("offline qr stream text codec round-trips", () => {
  const payload = buildPayload(128);
  const encoded = encodeQrFrameText(payload, OfflineQrStreamFrameEncoding.base64);
  const decoded = decodeQrFrameText(encoded, OfflineQrStreamFrameEncoding.base64);
  assert.deepEqual(decoded, payload);
});

test("offline qr stream scan loop ingests frames", async () => {
  const payload = buildPayload(512);
  const frames = OfflineQrStreamEncoder.encodeFrameBytes(payload, { chunkSize: 200 });
  const encodedFrames = frames.map((frame) =>
    encodeQrFrameText(frame, OfflineQrStreamFrameEncoding.base64),
  );
  const result = await scanQrStreamFrames(encodedFrames, {
    frameEncoding: OfflineQrStreamFrameEncoding.base64,
  });
  assert.ok(result);
  assert.equal(result.isComplete, true);
  assert.deepEqual(result.payload, payload);
});

test("offline qr stream scan session ingests text frames", () => {
  const payload = buildPayload(256);
  const frames = OfflineQrStreamEncoder.encodeFrameBytes(payload, { chunkSize: 200 });
  const session = new OfflineQrStreamScanSession({
    frameEncoding: OfflineQrStreamFrameEncoding.base64,
  });
  let result = null;
  for (const frame of frames) {
    const encoded = encodeQrFrameText(frame, OfflineQrStreamFrameEncoding.base64);
    result = session.ingest(encoded);
  }
  assert.ok(result);
  assert.equal(result.isComplete, true);
  assert.deepEqual(result.payload, payload);
});

test("offline qr stream fixtures round-trip", () => {
  const fixturePath = new URL("../../../fixtures/qr_stream/qr_stream_basic.json", import.meta.url);
  const fixture = JSON.parse(fs.readFileSync(fixturePath, "utf8"));
  const payload = Buffer.from(fixture.payload_hex, "hex");
  const frames = fixture.frames.map((frame) => Buffer.from(frame.bytes_hex, "hex"));
  const decoder = new OfflineQrStreamDecoder();
  let result = null;
  for (const frame of frames) {
    result = decoder.ingest(frame);
  }
  assert.ok(result);
  assert.equal(result.isComplete, true);
  assert.deepEqual(result.payload, payload);
});

test("offline qr stream sakura-storm skin matches preset", () => {
  assert.equal(sakuraStormQrStreamTheme.name, "sakura-storm");
  assert.deepEqual(sakuraStormQrStreamTheme.backgroundStart, {
    red: 0.05,
    green: 0.02,
    blue: 0.08,
  });
  assert.equal(sakuraStormQrStreamSkin.name, "sakura-storm");
  assert.equal(sakuraStormQrStreamSkin.frameRate, 12);
  assert.equal(sakuraStormQrStreamSkin.petalDriftSpeed, 0.6);
  assert.equal(sakuraStormQrStreamSkin.progressOverlayAlpha, 0.34);
});

test("offline qr stream sakura-storm preset recovers dropped frame in scan loop", () => {
  const payload = buildPayload(6 * 1024);
  const frames = OfflineQrStreamEncoder.encodeFrames(payload, {
    chunkSize: 336,
    parityGroup: 4,
  });
  const header = frames.find((frame) => frame.kind === OfflineQrStreamFrameKind.header);
  const dataFrames = frames.filter((frame) => frame.kind === OfflineQrStreamFrameKind.data);
  const parityFrames = frames.filter((frame) => frame.kind === OfflineQrStreamFrameKind.parity);
  assert.ok(header);
  assert.equal(dataFrames.length > 0, true);
  assert.equal(parityFrames.length > 0, true);

  const session = new OfflineQrStreamScanSession({
    frameEncoding: OfflineQrStreamFrameEncoding.base64,
  });
  session.ingest(encodeQrFrameText(header.encode(), OfflineQrStreamFrameEncoding.base64));
  let result = null;
  for (const frame of dataFrames) {
    if (frame.index === 1) {
      continue;
    }
    result = session.ingest(
      encodeQrFrameText(frame.encode(), OfflineQrStreamFrameEncoding.base64),
    );
  }
  for (const frame of parityFrames) {
    result = session.ingest(
      encodeQrFrameText(frame.encode(), OfflineQrStreamFrameEncoding.base64),
    );
  }
  assert.ok(result);
  assert.equal(result.isComplete, true);
  assert.deepEqual(result.payload, payload);
  assert.equal(result.recoveredChunks, 1);
});
